[Writer] Add Privy signer migration guides (React + JWT auth) (#1230)
alchemyplatform/docs: content/wallets/wallet-integrations/privy/react-migration.mdx

---

> title: "Migrate embedded wallets from Alchemy to Privy (React)"

Reviewer: the title kinda makes me thing that i'm going to be totally replacing alchemy w/ privy, instead of just migrating the signer

Author: Fixed: Retitled the guide to "Migrate the Alchemy signer to Privy (React)" and reworded the intro to emphasize that only the signer changes — Alchemy's transaction infrastructure and gas sponsorship stay in place. Commit: 3d7b2b20

After completing this migration, users will log in with Privy (wallet addresses are preserved, assets are maintained) and transactions will continue to be sent and sponsored by Alchemy.

<Info>If you authenticate users with JWT (JSON Web Token) authentication instead of standard auth methods, use the [JWT auth migration guide](/wallets/wallet-integrations/privy/jwt-auth-migration) instead.</Info>

## Overview

1. **Set up Privy and swap your app from Alchemy auth** — Replace `@account-kit/react` with `@privy-io/react-auth`
2. **Reconnect sending and gas sponsorship** — Wire Privy wallets to Alchemy's transaction infrastructure so your app works end-to-end again
3. **Install the migration SDK** — Drop-in UI that migrates wallet keys when users log in
4. **Export users from Alchemy and import into Privy** — Transfer user data and auth methods (this is not the transfer of keys)

## Step 1: Set up Privy and replace Alchemy auth

Before anything else, you need to replace Alchemy's auth SDK with Privy's SDK to support login and sign up.

1. Go to the [Privy Dashboard](https://dashboard.privy.io/) and set up an account
2. Follow the steps to set up a Privy application using `react-auth`

> 3. Under **Login methods**, enable the same auth methods your Alchemy users signed up with (email, Google, passkeys, etc.)

Reviewer: ```suggestion
2. Under **User Management - Authentication**, enable the same auth methods your Alchemy users signed up with (email, Google, passkeys, etc.)
```

Reviewer: step 2 feels redundant & confuses me

Reviewer: also i think the renamed this page on the nav

<img width="1129" height="946" alt="Screenshot 2026-04-20 at 4 35 39 PM" src="https://github.com/user-attachments/assets/55cc7fbd-7e3d-4c4a-ac1a-947bd1b8efe7" />

Author: Fixed: Applied your suggestion — updated the step to reference **User Management - Authentication** and listed the auth method examples. Commit: 3d7b2b20

Author: Fixed: Agreed — the old step 2 ("Follow the steps to set up a Privy application using react-auth") was just pointing vaguely at Privy's own setup. Removed it and condensed down to create-app + enable-auth-methods. Commit: 3d7b2b20

Author: Fixed: Trusting your SME call here — updated the section name to **User Management - Authentication** to match the current Privy dashboard nav. Commit: 3d7b2b20

### 1a. Install Privy and remove Alchemy auth packages

```bash
npm install @privy-io/react-auth
npm uninstall @account-kit/react
```

### 1b. Replace the Alchemy provider with Privy

Remove `AlchemyAccountProvider` and wrap your app with `PrivyProvider` instead.

During the migration window, set `createOnLogin` to `"off"` so Privy doesn't create a fresh wallet before the migration SDK can import the existing one. Once all users have migrated, switch it back to `"users-without-wallets"`.

```tsx
import { PrivyProvider } from '@privy-io/react-auth';

function Providers({ children }) {
  return (
    <PrivyProvider
      appId={process.env.NEXT_PUBLIC_PRIVY_APP_ID}
      config={{
        embeddedWallets: {
          ethereum: {
            createOnLogin: "off",
          }
        },
      }}
    >
      {children}
    </PrivyProvider>
  );
}
```

### 1c. Replace Alchemy hooks with Privy equivalents

| Before (Alchemy) | After (Privy) |
|---|---|
| `useSignerStatus().isConnected` | `usePrivy().authenticated` |
| `useAuthModal().openAuthModal()` | `usePrivy().login()` |

> | `useLogout()` | `usePrivy().logout` |

Reviewer: ```suggestion
| `useLogout()` | `usePrivy().logout()` |
```

consistency

Author: Fixed: Added the parens — `usePrivy().logout()` now matches the other rows in the table. Commit: 3d7b2b20

| `useUser()` | `usePrivy().user` — Email is at `user.email?.address`, not `user.email` |
| `useSmartAccountClient()` | See [Step 2](#step-2-reconnect-sending-and-gas-sponsorship) below — uses `@alchemy/wallet-apis` |
| `useSendUserOperation()` | `client.sendCalls()` — See [Step 2](#step-2-reconnect-sending-and-gas-sponsorship) below |

Update every file that imports from `@account-kit/react`.

At this point: Your app should compile, users can log in with Privy, but sending transactions is broken because the old Alchemy smart account hooks are gone. Step 2 fixes this.

## Step 2: Reconnect sending and gas sponsorship

Wire Privy wallets to Alchemy's transaction infrastructure so gasless transactions, batching, and all your existing flows work again. This uses Alchemy SDK v5 with Wallet APIs.

* The client defaults to EIP-7702, which delegates the Privy wallet to a smart wallet at send time.

> * If you were previously using ERC-4337 (with assets directly in smart accounts), see the EIP-7702 guide for non-7702 mode. You'll need to add an additional call to `wallet_requestAccount` before sending.

Reviewer: we should actually link to non-7702 guide here

Author: Fixed: Replaced the generic "EIP-7702 guide for non-7702 mode" link with a direct link to the non-7702 mode section: [/wallets/transactions/using-eip-7702#how-to-use-non-7702-mode](/wallets/transactions/using-eip-7702#how-to-use-non-7702-mode). Commit: 3d7b2b20

<Info>
For advanced setups, see the full [Privy signer integration guide](/wallets/third-party/signers/privy) or the [Alchemy Wallet APIs overview](/wallets).
</Info>

> ### Prerequisites

Reviewer: nit: feels weird to have "Prerequisites" when i'm 1/3 of the way through the guide already

Author: Fixed: Moved the Prerequisites section up to the top of the guide (right after the intro callout) so readers see the API key / Gas Manager policy requirements before they start. Also added "A React app currently using Alchemy Account Kit" as a prereq. The old `### Prerequisites` heading inside Step 2 has been removed. Commit: 3d7b2b20

import { useEffect, useState } from "react";
import type { LocalAccount } from "viem";

const usePrivySigner = () => {
  const { wallets: [wallet] } = useWallets();
  const [signer, setSigner] = useState<LocalAccount>();

  useEffect(() => {
    if (!wallet || signer) return;
    toViemAccount({ wallet }).then(setSigner);
  }, [wallet, signer]);
  return signer;
};
```

`useWallets()` returns the user's Privy wallets. `toViemAccount` converts the Privy wallet into a standard viem `LocalAccount`, which is what the Alchemy wallet client needs as a signer. The signer will be `undefined` until the wallet is ready — your UI should handle that loading state.

### 2c. Send gasless transactions

```tsx
import { useMemo, useCallback } from "react";
import { zeroAddress } from "viem";
import { createSmartWalletClient, alchemyWalletTransport } from "@alchemy/wallet-apis";
import { arbitrumSepolia } from "viem/chains";

function SendTransaction({ signer }: { signer: LocalAccount }) {
  const client = useMemo(
    () =>
      createSmartWalletClient({
        signer,
        transport: alchemyWalletTransport({
          apiKey: "YOUR_ALCHEMY_API_KEY",
        }),

>         chain: arbitrumSepolia,

Reviewer: probably obvious, but should we flag to use any chain here imported from viem?

Author: Fixed: Added a line above the snippet — "You can use any chain here — imported from `viem/chains`. Use whichever chain matches your Alchemy app configuration." — and added an inline comment on the import itself (`// Swap for any viem chain you support`). Commit: 3d7b2b20

Reviewer 2: Clarify that you should import the needed chain(s) from viem/chains. ensure the chain is enabled on your Alchemy app and support gas sponsorship.

Author: Done in feee7675. Reworded the callout to cover all three: import chain(s) from `viem/chains`, ensure the chain is enabled on your Alchemy app, and confirm it supports Gas Manager sponsorship.

## Step 3: Install the React migration SDK

The migration SDK is a drop-in React component that automatically detects users who need wallet key migration and walks them through re-authenticating with Alchemy to securely transfer their private keys to Privy.

### How it works

When a user logs in, the SDK:
1. **Detects migration need** — Checks if the user has Alchemy migration metadata and is missing embedded wallets
2. **Shows a migration modal** — Prompts the user to re-authenticate with their original Alchemy login method
3. **Migrates wallets** — Exports private keys from Alchemy's TEE and imports them into Privy's TEE via end-to-end encryption
4. **Confirms completion** — Shows a success screen and auto-closes

Users keep their exact same wallet addresses. No seed phrases, no manual steps.

### Supported authentication methods

The SDK supports migrating users who signed up with:
* Email (OTP)
* Google
* Twitter/X
* GitHub
* Discord
* Passkey (both anonymous and non-anonymous, e.g. attached an email)

> For non-React frameworks or other authentication methods, please [reach out](mailto:support@alchemy.com) for support.

Reviewer: i don't think anyone will have made it this far in the guide if they aren't using react. should we say this earlier?

Author: Fixed: Added "This guide is for **React** apps" to the intro callout at the top of the page, and also reworded the existing "For non-React frameworks" note in Step 3 so it no longer implies non-React is supported here. The title now also says "(React)" which helps. Commit: 3d7b2b20

### 3a. Install

```bash

> npm install @privy-io/alchemy-migration @account-kit/react @account-kit/infra

Reviewer: do we need to install `@account-kit/react` here? we just removed it in step 2.

Reviewer: and i think they should already have the infra pkg, assuming it's not a sub-dep of the migration pkg.

Author: Fixed: Split the install step. The migration SDK install is now just `npm install @privy-io/alchemy-migration` on its own. I added a short note right after explaining that if `@account-kit/react` or `@account-kit/infra` aren't already in the project (which for `@account-kit/react` they won't be after step 1a), install them as a follow-up. That way we're not re-installing what was just removed without an explanation. Commit: 3d7b2b20

Author: Fixed: Same edit as the comment above handles this too. `@account-kit/infra` is no longer in the unconditional install line — the follow-up note says to only install it if it isn't already resolved in the project. If it's a sub-dep of the migration pkg or still present from the Alchemy setup, users can skip that second install. Commit: 3d7b2b20

Reviewer 2: Remove the note to install @account-kit/react. We only need to install the migration pakage.

Author: Done in feee7675. Removed the follow-up `npm install @account-kit/react @account-kit/infra` note. Step 3a is now just `npm install @privy-io/alchemy-migration`.

```

### 3b. Import styles

```tsx
import '@privy-io/alchemy-migration/styles.css';
```

### 3c. Create the Alchemy config

Use `createMigrationConfig` to create an Alchemy config optimized for the migration flow. It handles session management, OAuth redirect mode, and stale session cleanup automatically.

The `auth.sections` should include all the login methods originally used on Alchemy. The SDK will automatically detect which method each user needs and present the right sign-in flow.

<Note>This is the same even if you used custom UI with `useAuthenticate` — follow the same config below and add all of your used auth methods. Only pre-built components are supported for the migration flow.</Note>

```tsx
import { createMigrationConfig } from '@privy-io/alchemy-migration';
import { alchemy } from '@account-kit/infra';

> import { sepolia } from '@account-kit/infra';

Reviewer: these 2 imports can be 1 line

Author: Fixed: Combined into `import { alchemy, sepolia } from '@account-kit/infra';`. Commit: 3d7b2b20

const alchemyConfig = createMigrationConfig(
  {
    transport: alchemy({ apiKey: process.env.NEXT_PUBLIC_ALCHEMY_API_KEY! }),
    chain: sepolia, // Use your app's chain
    ssr: true,
  },
  {
    auth: {
      sections: [
        [{ type: 'email' }],
        [
          { type: 'passkey' },
          { type: 'social', authProviderId: 'google', mode: 'redirect' },
          // Add any other providers your users signed up with
        ],
      ],
    },
  },
);
```

> <Warning>Never commit your Alchemy API key to source control. Load it from an environment variable (for example `NEXT_PUBLIC_ALCHEMY_API_KEY` in Next.js) and restrict the key to the correct apps, networks, and origins from the [Alchemy Dashboard](https://dashboard.alchemy.com/).</Warning>

Reviewer: technically it's best for the api key to be secured on the backend

Author: Fixed: Strengthened the Warning — it now explicitly notes that for production apps the best practice is to keep the API key on the backend and proxy Alchemy requests through it rather than exposing it in the client bundle at all. Commit: 3d7b2b20

      <MigrationProvider
        alchemyConfig={alchemyConfig}
        privyAppId="YOUR_PRIVY_APP_ID"
        privyClientId="YOUR_PRIVY_APP_CLIENT_ID"
      >
        {children}
      </MigrationProvider>
    </PrivyProvider>
  );
}
```

The SDK automatically detects users who need migration and guides them through the key transfer process.

### MigrationProvider props

| Prop | Type | Required | Description |
|------|------|----------|-------------|
| `alchemyConfig` | `AlchemyAccountsConfigWithUI` | Yes | Config created via `createMigrationConfig` |
| `privyAppId` | `string` | Yes | Your Privy application ID |
| `privyClientId` | `string` | Yes | Your Privy app client ID |
| `showDebugButton` | `boolean` | No | Show debug buttons for testing (dev only) |

### After migration is complete

Once all users have migrated their wallets (e.g. logged in at least once):
* You can safely remove `@privy-io/alchemy-migration`, `@account-kit/react`, and `@account-kit/infra` from your app
* Set `createOnLogin` to `"users-without-wallets"` so new users get wallets automatically
* Remove the `MigrationProvider` wrapper

## Step 4: Deploy your updated app, export users, and import into Privy

> In order for Privy to be able to detect if a user needs migration, you need to migrate over user data including auth methods to Privy. This is not a migration of keys.

Reviewer: style nit: 

```suggestion
In order for Privy to be able to detect if a user needs migration, you need to migrate user data (including auth methods) to Privy. This is a separate step from the migration of keys.
```

Author: Fixed: Applied your style suggestion verbatim. Commit: 3d7b2b20

### Recommended cutover sequence

<Warning>
**Critical sequencing:** The user export from Alchemy must happen at the exact same time you deploy your updated app, or immediately after. If you export before deploying, users who sign up between the export and the deploy will be missing from Privy and won't get migrated. The sequence is: **Deploy → Export → Import** (or do them simultaneously).
</Warning>

1. Pause all new sign ups / pause app momentarily
2. Deploy your updated app with Privy auth
3. Immediately export from Alchemy
4. Import into Privy

> 5. Redeploy app

Reviewer: i think step 2 and 5 is the same? don't think it needs to be there twice.

Author: Fixed: Good catch — step 2 ("Deploy your updated app with Privy auth") already covers the deploy, and step 5 ("Redeploy app") was redundant with the "resume from pause" implied by the earlier pause step. Removed step 5; the sequence is now Pause → Deploy → Export → Import. Commit: 3d7b2b20

2. Navigate to **Wallets > Export** tab
3. Click **Export Users** to download a JSON file containing all users and their wallet addresses

### 4b. Import users into Privy

1. Go to the [Privy Dashboard](https://dashboard.privy.io/) and navigate to the relevant new app
2. Navigate to **User management > Users > Import users**
3. Drag and drop the exported JSON file from step 4a

Privy will create accounts for each user with the appropriate linked accounts and migration metadata. The migration SDK uses this metadata to detect which users need wallet migration.

## FAQ

<AccordionGroup>
  <Accordion title="Do users need to do anything?">
    They just sign in once with their original Alchemy login method. The SDK handles everything.
  </Accordion>

  <Accordion title="Will wallet addresses change?">
    No. Wallet addresses are preserved. Only the signer is moved over to Privy.
  </Accordion>

  <Accordion title="Are private keys ever exposed?">
    No. Keys are exported from Alchemy's TEE and imported into Privy's TEE using end-to-end encryption. Key material is never exposed in plaintext.
  </Accordion>

  <Accordion title="What if a user has both ETH and SOL wallets?">

>     Both are migrated automatically. The SDK migrates each chain independently.

Reviewer: ```suggestion
    Both are migrated automatically. The SDK migrates each wallet independently.
```

saying "chain" here is confusing to me

Author: Fixed: Applied your suggestion — "Both are migrated automatically. The SDK migrates each wallet independently." Agreed that "chain" was misleading here since the FAQ is specifically about a user having both ETH and SOL wallets. Commit: 3d7b2b20

---
title: "Migrate the Alchemy signer to Privy (React)"
description: "Step-by-step guide to migrate your React app's signer from Alchemy Account Kit to Privy, while keeping Alchemy's transaction infrastructure and gas sponsorship."
slug: wallets/wallet-integrations/privy/react-migration
---

This guide walks you through migrating the **signer** for your app and users' embedded wallets from Alchemy's Account Kit to Privy. You'll keep Alchemy's transaction infrastructure and gas sponsorship — only the signer (auth + key custody) changes. The migration SDK handles re-authenticating users with Alchemy, securely exporting their private keys, and importing them into Privy.

After completing this migration, users will log in with Privy (wallet addresses are preserved, assets are maintained) and transactions will continue to be sent and sponsored by Alchemy.

> <Info>This guide is for **React** apps. For non-React frameworks, [reach out](mailto:support@alchemy.com) for support. If you authenticate users with JWT (JSON Web Token) authentication instead of standard auth methods, use the [JWT auth migration guide](/wallets/wallet-integrations/privy/jwt-auth-migration) instead.</Info>

Reviewer 2: Also we should add notes to the top of each migration guide that points to th overview so developers ensure they are using the correct account type and connection to avoid problems

Author: Done in 33ef77f8. New section order is: intro callout → What to expect → Prerequisites → Edge cases.

## Overview

1. **Set up Privy and swap your app from Alchemy auth** — Replace `@account-kit/react` with `@privy-io/react-auth`
2. **Reconnect sending and gas sponsorship** — Wire Privy wallets to Alchemy's transaction infrastructure so your app works end-to-end again
3. **Install the migration SDK** — Drop-in UI that migrates wallet keys when users log in
4. **Export users from Alchemy and import into Privy** — Transfer user data and auth methods (this is not the transfer of keys)

## Step 1: Set up Privy and replace Alchemy auth

Before anything else, you need to replace Alchemy's auth SDK with Privy's SDK to support login and sign up.

1. Go to the [Privy Dashboard](https://dashboard.privy.io/) and set up an account
2. Follow the steps to set up a Privy application using `react-auth`
3. Under **Login methods**, enable the same auth methods your Alchemy users signed up with (email, Google, passkeys, etc.)

### 1a. Install Privy and remove Alchemy auth packages

```bash
npm install @privy-io/react-auth
npm uninstall @account-kit/react
```

### 1b. Replace the Alchemy provider with Privy

> Remove `AlchemyAccountProvider` (and `cookieToInitialState` if using SSR) and wrap your app with `PrivyProvider` instead.

Reviewer 3: should we omit this cookieToInitialState thing?

Author: Applied — dropped the `cookieToInitialState` mention. Simpler to just tell people to remove `AlchemyAccountProvider` and move on.

> Ensure you set `createOnLogin` to `"user-without-wallet"`, and that the plugin for wallet creation is set as follows:

Author: Fixed — updated the prose on line 39 to use `"users-without-wallets"` to match the code examples. The code samples were already correct per the source guide.

```tsx
import { PrivyProvider, createWalletCreationOnLoginPlugin, User } from '@privy-io/react-auth';

function Providers({ children }) {
  // Add custom logic to only create a new embedded wallet
  const walletCreationPluginOptions = {
    shouldCreateWallet: ({user}: {user: User}) =>

>       user.customMetadata?.['alchemy_org_id'] === undefined,

Reviewer 3: was this confirmed with danny? I remember they were looking to ship something that just did this automatically

Author: Good catch — went with the simpler path that matches the FAQ at the bottom: drop the custom wallet creation plugin entirely and set `createOnLogin: "off"` during the migration window, then flip it to `"users-without-wallets"` once everyone's migrated. That way we're not relying on the `alchemy_org_id` metadata check, and it's consistent whether or not Privy ships auto-detection down the road.

Reviewer 2: @JackReacher0807 Revert this back to the original implementation. We do need the custom wallet creation plugin and should not pass the createOnLogin. Auto-detection was not shipped.

Author: Reverted in feee7675. Step 1b is back to the original implementation: `createWalletCreationOnLoginPlugin` with `shouldCreateWallet` gated on `alchemy_org_id`, and `createOnLogin: "users-without-wallets"`. Dropped the auto-detection / `createOnLogin: "off"` framing.

  };

  return (
    <PrivyProvider
      appId={process.env.NEXT_PUBLIC_PRIVY_APP_ID}
      config={{
        plugins: [createWalletCreationOnLoginPlugin(walletCreationPluginOptions)],
        embeddedWallets: {
          ethereum: {
            createOnLogin: "users-without-wallets",
          }
        },
      }}
    >
      {children}
    </PrivyProvider>
  );
}
```

### 1c. Replace Alchemy hooks with Privy equivalents

| Before (Alchemy) | After (Privy) |
|---|---|
| `useSignerStatus().isConnected` | `usePrivy().authenticated` |
| `useAuthModal().openAuthModal()` | `usePrivy().login()` |

> | `useLogout()` | `usePrivy().logout` — Function, not hook |

Reviewer 3: isnt the above alos a function? why are we specifying?

Author: You're right — `useAuthModal().openAuthModal()` is also a function returned from a hook, so singling out `useLogout()` was misleading. Dropped the note.

| `useUser()` | `usePrivy().user` — Email is at `user.email?.address`, not `user.email` |
| `useSmartAccountClient()` | See [Step 2](#step-2-reconnect-sending-and-gas-sponsorship) below — uses `@alchemy/wallet-apis` |
| `useSendUserOperation()` | `client.sendCalls()` — See [Step 2](#step-2-reconnect-sending-and-gas-sponsorship) below |

> | `cookieToInitialState` (SSR) | Remove entirely |

Reviewer 3: do we need this? I think it's p obvious you have to get rid of that. it would be unused if you remove the provider

Author: Agreed, redundant once the provider section no longer mentions it. Removed the row.

Update every file that imports from `@account-kit/react`.

At this point: Your app should compile, users can log in with Privy, but sending transactions is broken because the old Alchemy smart account hooks are gone. Step 2 fixes this.

## Step 2: Reconnect sending and gas sponsorship

Wire Privy wallets to Alchemy's transaction infrastructure so gasless transactions, batching, and all your existing flows work again. This uses Alchemy SDK v5 with Wallet APIs.

<Info>

> Full guide for Privy + Alchemy: [Privy signer integration](/wallets/third-party/signers/privy). More on Alchemy SDK v5: [Wallet APIs overview](/wallets).

Reviewer 3: this confused be a little bit. should we move it to tyhe bottom or explicitly list it as for more advnaced migrations check out this? I was torn whether or not I had to follow this link vs what's written here

Author: Moved the `<Info>` callout below the 7702/4337 bullets and reframed it as 'for advanced setups, see…'. That way the inline steps read as the canonical path and the link is clearly optional.

Wire Privy wallets to Alchemy's transaction infrastructure so gasless transactions, batching, and all your existing flows work again. This uses Alchemy SDK v5 with Wallet APIs.

* The client defaults to EIP-7702, which delegates the Privy wallet to a smart wallet at send time.

> * If you were previously using ERC-4337 (with assets directly in smart accounts), see the [non-7702 mode guide](/wallets/transactions/using-eip-7702#how-to-use-non-7702-mode). You'll need to add an additional call to `wallet_requestAccount` before sending.

Reviewer 2: Add a note here to continue to use low level SDK if you're not using MAv2 or different account type as well

Author: Done in 33ef77f8. The overview now defaults to the React path in the intro callout (labeled "the default path"), and JWT is called out as a special case. The What-to-expect steps describe only the React packages; JWT is referenced only via a pointer to the JWT sub-guide.

}
```

At this point: Your app should fully work with Privy auth + Alchemy gas sponsorship. Users can log in, sign, and send gasless transactions. Now you need to move existing users' wallet keys over.

## Step 3: Install the React migration SDK

The migration SDK is a drop-in React component that automatically detects users who need wallet key migration and walks them through re-authenticating with Alchemy to securely transfer their private keys to Privy.

### How it works

When a user logs in, the SDK:
1. **Detects migration need** — Checks if the user has Alchemy migration metadata and is missing embedded wallets
2. **Shows a migration modal** — Prompts the user to re-authenticate with their original Alchemy login method
3. **Migrates wallets** — Exports private keys from Alchemy's TEE and imports them into Privy's TEE via end-to-end encryption
4. **Confirms completion** — Shows a success screen and auto-closes

Users keep their exact same wallet addresses. No seed phrases, no manual steps.

### Supported authentication methods

The SDK supports migrating users who signed up with:
* Email (OTP)
* Google
* Twitter/X
* GitHub
* Discord

> * Passkey (non-anonymous, e.g. attached an email)

Reviewer 3: I think we support anonymous too? see comment on original doc

Author: Fixed — updated to 'both anonymous and non-anonymous' for passkeys.

For non-React frameworks or other authentication methods, please [reach out](mailto:support@alchemy.com) for support.

### 3a. Install

```bash
npm install @privy-io/alchemy-migration @account-kit/react @account-kit/infra
```

### 3b. Import styles

```tsx
import '@privy-io/alchemy-migration/styles.css';
```

### 3c. Create the Alchemy config

Use `createMigrationConfig` to create an Alchemy config optimized for the migration flow. It handles session management, OAuth redirect mode, and stale session cleanup automatically.

The `auth.sections` should include all the login methods originally used on Alchemy. The SDK will automatically detect which method each user needs and present the right sign-in flow.

<Note>This is the same even if you used custom UI with `useAuthenticate` — follow the same config below and add all of your used auth methods. Only pre-built components are supported for the migration flow.</Note>

```tsx
import { createMigrationConfig } from '@privy-io/alchemy-migration';
import { alchemy } from '@account-kit/infra';
import { sepolia } from '@account-kit/infra';

const alchemyConfig = createMigrationConfig(
  {

>     transport: alchemy({ apiKey: 'YOUR_ALCHEMY_API_KEY' }),

Reviewer 3: fyi alchemy key exposed. should we add a note?

Author: Good call. Kept the placeholder (`YOUR_ALCHEMY_API_KEY`) but changed the snippet to load it from an env var (`process.env.NEXT_PUBLIC_ALCHEMY_API_KEY`) and added a <Warning> right under the code block about not committing keys and restricting them in the dashboard.

    chain: sepolia, // Use your app's chain
    ssr: true,
  },
  {
    auth: {
      sections: [
        [{ type: 'email' }],
        [
          { type: 'passkey' },

>           { type: 'social', authProviderId: 'google', mode: 'popup' },

Reviewer 3: I think mode should always be redirect? it gets overwritten so doesnt matter but mode here doesn't do anything

Author: Switched to `mode: 'redirect'`. Yeah it gets overwritten for the migration flow, but no reason to leave a misleading value in the example.

          // Add any other providers your users signed up with
        ],
      ],
    },
  },
);
```

### 3d. Set up the provider

Wrap your app with `MigrationProvider` inside your existing `PrivyProvider`:

```tsx
import { PrivyProvider } from '@privy-io/react-auth';
import { MigrationProvider } from '@privy-io/alchemy-migration';

function App({ children }) {
  return (
    <PrivyProvider appId="YOUR_PRIVY_APP_ID">
      <MigrationProvider
        alchemyConfig={alchemyConfig}
        privyAppId="YOUR_PRIVY_APP_ID"
        privyClientId="YOUR_PRIVY_APP_CLIENT_ID"
      >
        {children}
      </MigrationProvider>
    </PrivyProvider>
  );
}
```

The SDK automatically detects users who need migration and guides them through the key transfer process.

### MigrationProvider props

| Prop | Type | Required | Description |
|------|------|----------|-------------|
| `alchemyConfig` | `AlchemyAccountsConfigWithUI` | Yes | Config created via `createMigrationConfig` |
| `privyAppId` | `string` | Yes | Your Privy application ID |
| `privyClientId` | `string` | Yes | Your Privy app client ID |
| `showDebugButton` | `boolean` | No | Show debug buttons for testing (dev only) |

> | `skipOrgVerification` | `boolean` | No | Skip Alchemy org ID verification (dev only) |

Reviewer 3: get rid of this

Author: Dropped `skipOrgVerification` from the prop table.
